Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs

```diff
@@ -0,0 +1,318 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.InputSystem;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Physics;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Teleport;
+using Microsoft.MixedReality.Toolkit.Internal.Utilities.Physics;
+using System;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX.Pointers
+{
+    public class TeleportPointer : LinePointer
+    {
+        [SerializeField]
+        private MixedRealityInputAction teleportAction = MixedRealityInputAction.None;
+
+        [SerializeField]
+        [Tooltip("The angle from the pointer's forward position that will activate the teleport.")]
+        private float activationAngle = 45f;
+
+        [SerializeField]
+        [Tooltip("If Pressing 'forward' on the thumbstick gives us an angle that doesn't quite feel like the forward direction, we apply this offset to make navigation feel more natural")]
+        private float angleOffset = 0f;
+
+        [SerializeField]
+        private float upDirectionThreshold = 0.2f;
+
+        [SerializeField]
+        protected Gradient LineColorHotSpot = new Gradient();
+
+        [SerializeField]
+        [Tooltip("Layers that are considered 'valid' for navigation")]
+        protected LayerMask ValidLayers = 1 << 0; // Default
+
+        [SerializeField]
+        [Tooltip("Layers that are considered 'invalid' for navigation")]
+        protected LayerMask InvalidLayers = 1 << 2; // Ignore raycast
```

Comment:
Doesnt Unity have a way to access these by name?

Reply:
yes, but they're expensive, and a simple bit flag is enough. I simply copied from master for this one and left as is.

Reply:
>I simply copied from master for this one and left as is.

Where did you copy from master for this?

Reply:
Nvm, couldn't find it. I think it was from HUX actually. I can't remember exactly where I got it.

Reply:
let's file a beta issue to read the layer values (by name) at startup and cache their mask values.... just in case they change in a future version of Unity

Reply:
> just in case they change in a future version of Unity

I highly doubt that, but sure.

Reply:
> layer values (by name) at startup

Sounds expensive.

Reply:
>Sounds expensive

One time at startup?

Reply:
#2564

Reply:
I think you guys are being a bit overboard here.

The inspector for this is user friendly and shows all the layers in the project in the dropdown.
![image](https://user-images.githubusercontent.com/13334553/43790893-d0426fe4-9a28-11e8-9355-acebef0e241f.png)

Reply:
The only thing I've done here is set it to the default and ignore raycast layers which will most likely never change (nor have they in 10 years that unity has used them)

Reply:
They're even blocked out from being editable:
![image](https://user-images.githubusercontent.com/13334553/43791077-5bccf890-9a29-11e8-84b1-1b573fa039c6.png)

Reply:
I found a nice const value for a substitute 👍 

`Physics.DefaultRaycastLayers` and `Physics.IgnoreRaycastLayer`